### ISSUE TITLE
Level matching via ReplayGain

### ISSUE THREAD

**Issue by zrav** [airsonic/airsonic#525]
Currently, Sub/Libre/Airsonic lack a volume level matching feature to avoid the large jumps in volume of different recordings. I tested the EBU R128 implementation in recent versions of ffmpeg, and it works okay most of the time, but it's not a replacement for a solution like ReplayGain, that allows you to scan your media library once, storing the leveling metadata in the files, which are then applied during playback by a supporting player. I'd like to see this in Airsonic. 

ffmpeg currently only supports reading the RG tags, not applying them during transcode, so we'd need to implement some functionality in Airsonic to get this working. Airsonic needs to provide the RG gain value as a variable for the transcode command, so it can be used as parameter for a ffmpeg volume filter. I don't know what the best way to get this data is, if during library scan (storing it in the DB) or just-in-time before playback starts.

The RG tag fields are:
- replaygain_album_gain
- replaygain_album_peak
- replaygain_track_gain
- replaygain_track_peak

More details about the tag format and interpretation can be found here: http://wiki.hydrogenaud.io/index.php?title=ReplayGain_1.0_specification 

As for configuration, the feature would need to have the states "off", "album mode" and "track mode", as well as a checkbox for toggling clipping prevention. Additionally, a field for defining a default gain for tracks lacking RG data would be great.

**Comment by muff1nman** [comment]
This would be a great feature. Thanks for the well written feature request.

**Comment by texneus** [comment]
The replaygain tags are passed as part of the metadata to the player, in that way the player (i.e. DSub) can apply the necessary gain.  If you want to do it as part of the transcode rather than as part of playback, the _volume_ function of FFMpeg may be of interest:  [https://trac.ffmpeg.org/wiki/AudioVolume](url)

The following thread provides a means for implimentation.  I did this a while back before I found DSub and it does work, but takes a bit of effort.  Maybe Airsonic can provide the necessary tags as a variable that can be input direct into the FFMpeg transcode parameters?

[http://forum.madsonic.org/viewtopic.php?f=29&t=1097](url)

**Comment by f10org** [comment]
Having the client player apply volume leveling based on the RG data would be the best solution, as it would work for both the cases where the files are streamed without transcoding, as well as with (ffmpeg does preserve the RG tags).
The question is whether the MediaElement.js project is interested in this feature...

**Comment by eharris** [comment]
I've been playing with doing transcoding via a script for volume (ReplayGain) normalization, and while it does work, it certainly has some drawbacks:

* There's no way to control whether the gain value used is per-track or per-album, as that depends on the player knowing how things are being played and selecting the appropriate gain level.
* Requires customized transcoding/downsampling settings, and adds to the complexity of the airsonic config and environment
* Since the transcoding settings apply globally, there's no way to avoid transcoding for gain adjustment for users/players that do not need it (i.e. for ones that are already capable of doing gain adjustment on their own)
* How do you handle files that don't already contain gain info in tags?
* And probably most importantly, it subjects you to another generation of quality loss (very undesirable when already using lossy formats, IMHO)

There are some tricks I've considered to get around some of these issues.  For instance, the problem of not being able to control which users/players get transcoding *could* be done at the user level by using a special low "Max bitrate" value in the config for that user (like 32k) that your script uses to know not to actually do the transcode and just pass the data through unmolested.  But that has the side effect that now you can't also apply a real limit to that user's bandwidth usage.

A potential way to mitigate the transcoding loss of quality would be (as an example for mp3's) to have the transcoding script make a copy of the original source file to some temporary storage area, apply mp3gain adjustments to that copy (which doesn't require transcoding) and then stream the data from that modified file.  This works if the file isn't too large (otherwise the delay before it starts to play gets nasty), and if you have a good way to manage cleaning up stale temp files.

To me, it seems like there are a number of steps that could be made to improve the situation, with different levels of impact to the codebase (given here in approximate order of increasing impact/complexity):

* Add a new variable substitution to the transcoding config that allows you to pass the username (%u?) to the transcoding script.  This allows you to handle different settings for different users entirely outside the airsonic config.
* Add a new flag to the config of each User/Player that specifies if it should get server-side gain adjustment, and a new variable substitution (%g?) to pass it to the transcoding commands.  If enabled, this should also override the the bitrate limit conditions and always call the transcode/downsample commands even if the file wouldn't normally need to be transcoded/downsampled.
* Add saving of gain tag info to the airsonic db when media files are scanned, and a new variable to pass the indicated gain level to the transcode commands.  This allows direct support in the transcoding commands without requiring a separate script and avoids the scan of the file to find the gain level info, but depends on the gain info already being present in tags.  And still doesn't solve the per-track/per-album gain issue.
* Add support for built-in client side gain adjustment to the web player.  This only works if the tags holding the gain adjustments are stored at the beginning of the media file, unless support is also added for storing them in the airsonic db and passing them to the player somehow.  Depending on how it is done, this could be the first option to give us some level of smart per-album or per-track gain adjustment, at least in the web player.

Of course the best option all around would be for the client-side players to have built-in support of doing gain adjustment at playback.  But even that may have some issues, for instance, if the gain info is stored in APE tags which are usually at the end of the file, the player doesn't easily have access to the gain info when streaming the file unless it does some nasty seek trickery.  Would it be better to have airsonic pass the gain settings to the player some other way, or would that require API changes that would have the potential to break existing clients?

**Comment by phirestalker** [comment]
according to this link: https://video.stackexchange.com/questions/19527/how-to-modify-audio-level-of-video-according-to-replaygain-info

FFmpeg can directly use the replaygain tags to normalize the output stream with a line such as`ffmpeg -i input.mp4 -c:v copy -af volume=replaygain=track output.mp4`

someone here said FFmpeg couldn't directly use the tag values to alter the volume of the audio stream, but if it is an audio filter, then it should work that way? or am I missing something? Also there are some devices that could never do replaygain, because they are "too good for it". Like google home units and chromecasts for instance. Google can't even normalize their own music from their streaming music service, so transcoding the files with replaygain on the fly before sending to dumb players would be awesome.

I agree that per-user settings would work good

**Comment by zrav** [comment]
You are correct, ffmpeg does allow for applying RG volume change (incidentally I just found out it's currently borked for mpc input files). The downside is of course that it requires transcoding for it to be applied. I tried configuring airsonic to transcode all formats to flac, so it is lossless at least, however I then found that seeking in the web player is not possible with flac...

**Comment by phirestalker** [comment]
have you tried ALAC (Apple Lossless Audio Codec)? Maybe the encapsulating MP4 container will allow it to seek?

I also would like to do it without transcoding. I guess if the files are already lossless it doesn't matter.  I have been trying to get my collection again in lossless format.

**Comment by zrav** [comment]
ALAC is only supported by Safari.
I tried FLAC in MP4 via `ffmpeg -i %s -map 0:0 -c:a flac -af volume=replaygain=album -v 0 -f mp4 -strict -2 -` and although Chrome and FireFox should support it (https://developer.mozilla.org/en-US/docs/Web/HTML/Supported_media_formats), playback via the webplayer is not possible, unfortunately.

**Comment by ae-g-i-s** [comment]
@zrav Have you tried it with using the regular FLAC format or oga? FLAC in mp4 is a) perverse, b) only supported in `<video>` elements according to the page you linked (whereas FLAC in oga or its own container is supported in `<audio>` elements by both FF and Chrome).
Of course, streaming FLAC is highly questionable in terms of bandwidth, though I understand your transcoding concerns.

**Comment by zrav** [comment]
@ae-g-i-s  Straight FLAC plays, but is not seekable. FLAC in OGG, MP4 or MKA doesn't play at all, although I might be missing some necessary configuration. Maybe someone can confirm?

**Comment by dest4** [comment]
I would love to see this feature implemented.

Transcoding all files to apply gain seems a suboptimal approach in terms of CPU usage and/or storage.

Why not "simply" set the volume of the MediaElement to `gainVolume * userVolume`, where each volume is between 0 and 1? We do not need the collaboration of the MediaElement.js community here (cf https://github.com/airsonic/airsonic/issues/525#issuecomment-328733130 ).

**Comment by dest4** [comment]
I would love to see this feature implemented.

Transcoding all files to apply gain seems a suboptimal approach in terms of CPU usage and/or storage.

Why not "simply" set the volume of the MediaElement to `gainVolume * userVolume`, where each volume is between 0 and 1? We do not need the collaboration of the MediaElement.js community here (cf https://github.com/airsonic/airsonic/issues/525#issuecomment-328733130 ).

**Comment by eharris** [comment]
I've taken to using recent versions of ffmpeg with the `loudnorm` filter, which implements the EBU R128 algorithm (arguably better than ReplayGain).  For "perfect" listening, the two pass mode (one analysis pass followed by linear rather than dynamic/broadcast gain adjustment) is suggested, but for me, since I have a mixed audio library of many different formats (ogg, wav, flac, opus, mp3, aac, m4a) rather than having to continually try to consistently retag everything and deal with the nuances of different tags/formats, and for maximizing space/caching on my mobile devices) this is the easiest for for general listening.

So as an interim solution, I use the following for transcoding: `ffmpeg -i %s -map 0:0 -b:a %bk -v 0 -filter:a loudnorm=I=-20 -f m4a -`

To make this workaround really work well, all that is needed is a way to force transcoding of everything for specific users/players (the clients that don't support adjusting the replay gain themselves), rather than depending on a format difference or exceeding bitrates for the transcoding to trigger.

**Comment by jallbrit** [comment]
> To make this workaround really work well, all that is needed is a way to force transcoding of everything for specific users/players (the clients that don't support adjusting the replay gain themselves), rather than depending on a format difference or exceeding bitrates for the transcoding to trigger.

@eharris, is there a workaround that allows this transcoding to apply to all outgoing audio that normally wouldn't trigger a transcoding (e.g. an mp3 that does not exceed max bitrate)?

**Comment by seamusdemora** [comment]
@dest4 :  
> I would love to see this feature implemented.
> 
> Transcoding all files to apply gain seems a suboptimal approach in terms of CPU usage and/or storage.
> 
> Why not "simply" set the volume of the MediaElement to `gainVolume * userVolume`, where each volume is between 0 and 1? We do not need the collaboration of the MediaElement.js community here (cf [#525 (comment)](https://github.com/airsonic/airsonic/issues/525#issuecomment-328733130) ).

I just dropped out of the cloud [with this search](https://duckduckgo.com/?t=ffnt&q=ffmpeg+and+replay+gain&ia=web). I may be missing the point entirely, and I am speaking only for myself of course - but here's my take on it: 

Building volume normalization into the music files (i.e. *transcoding*) beforehand (i.e. prior to *"playtime"*) seems like a sound idea to me (no pun intended); i.e. using `ffmpeg` as [outlined here](https://video.stackexchange.com/questions/19527/how-to-modify-audio-level-of-video-according-to-replaygain-info). It would seem that doing all of the necessary transcoding in the library as a *"one-time effort"* would be much less CPU-intensive than doing it prior to each *"playtime"*. 

That said, your idea for `gainVolume * userVolume` strikes me as a good one assuming `gainVolume` is the volume encoded in the file via ReplayGain. Have I misunderstood your suggestion?

**Comment by dest4** [comment]
> @dest4 :
> 
> > I would love to see this feature implemented.
> > Transcoding all files to apply gain seems a suboptimal approach in terms of CPU usage and/or storage.
> > Why not "simply" set the volume of the MediaElement to `gainVolume * userVolume`, where each volume is between 0 and 1? We do not need the collaboration of the MediaElement.js community here (cf [#525 (comment)](https://github.com/airsonic/airsonic/issues/525#issuecomment-328733130) ).
> 
> I just dropped out of the cloud [with this search](https://duckduckgo.com/?t=ffnt&q=ffmpeg+and+replay+gain&ia=web). I may be missing the point entirely, and I am speaking only for myself of course - but here's my take on it:
> 
> Building volume normalization into the music files (i.e. _transcoding_) beforehand (i.e. prior to _"playtime"_) seems like a sound idea to me (no pun intended); i.e. using `ffmpeg` as [outlined here](https://video.stackexchange.com/questions/19527/how-to-modify-audio-level-of-video-according-to-replaygain-info). It would seem that doing all of the necessary transcoding in the library as a _"one-time effort"_ would be much less CPU-intensive than doing it prior to each _"playtime"_.
> 
> That said, your idea for `gainVolume * userVolume` strikes me as a good one assuming `gainVolume` is the volume encoded in the file via ReplayGain. Have I misunderstood your suggestion?

I am perfectly in sync with your message.

**Comment by seamusdemora** [comment]
@phirestalker : 
> according to this link: https://video.stackexchange.com/questions/19527/how-to-modify-audio-level-of-video-according-to-replaygain-info
> 
> FFmpeg can directly use the replaygain tags to normalize the output stream with a line such as`ffmpeg -i input.mp4 -c:v copy -af volume=replaygain=track output.mp4`

I am learning from reading this thread how **little** I know about this volume equalization! 

May I ask you: How do the `replaygain` tags come to be in the file? I gather these are generated and added to the file, but how is that done? Are these the `track_gain` and `track_peak` figures [referenced here?](https://ffmpeg.org/ffmpeg-all.html#replaygain)

**Comment by pagdot** [comment]
@seamusdemora 
> Building volume normalization into the music files (i.e. transcoding) beforehand (i.e. prior to "playtime") seems like a sound idea to me (no pun intended); i.e. using ffmpeg as outlined here. It would seem that doing all of the necessary transcoding in the library as a "one-time effort" would be much less CPU-intensive than doing it prior to each "playtime".

Reencoding the file with only changes in its volume would still cause quality losses. By storing the data in the ID3 metadata the original audio track doesn't get touched. I don't think changing the volume of the track during playback should cause overhead because the absolute volume only needs to get calculated once per track.

@seamusdemora 
> May I ask you: How do the replaygain tags come to be in the file? I gather these are generated and added to the file, but how is that done? Are these the track_gain and track_peak figures referenced here?

There are multiple tools out there like MP3Gain or FFMPEG generating the replay gain value. The value gets stored in the ID3 metadata of the music file. If you want to know more [Wikipedia](https://en.wikipedia.org/wiki/ReplayGain) is a good starting point.

**Comment by seamusdemora** [comment]
@pagdot 

> Reencoding the file with only changes in its volume would still cause quality losses. 

According to the Wikipedia article you referenced, for MP3 files at least, MP3Gain can reverse the changes it makes to change the volume. 

But I agree that metadata or tags are preferable to making changes that alter an original lossless file. For some (most?) file types other than MP3, the problem with tags/metadata seems to be that a player with enough "smarts" to use the tags is required. My *'use case'* is ~99% Sonos with FLAC-encoded media; the tags don't help me as Sonos doesn't/can't use them. At least that's the way it looks now. I'm just beginning to look for solutions, and learn about all of this... your comments have been very helpful. 

> > May I ask you: How do the replaygain tags come to be in the file? I gather these are generated and added to the file, but how is that done? Are these the track_gain and track_peak figures referenced here?
> 
> There are multiple tools out there like MP3Gain or FFMPEG generating the replay gain value. The value gets stored in the ID3 metadata of the music file. If you want to know more [Wikipedia](https://en.wikipedia.org/wiki/ReplayGain) is a good starting point.

Unfortunately, the Wikipedia article doesn't actually explain how to add the ReplayGain tags/metadata. That information may be in the [`ffmpeg` documentation](https://ffmpeg.org/ffmpeg-all.html#replaygain), but I find the documentation a bit hard to follow in places. I've learned [how to *generate* `track_gain` and `track_peak`](https://ffmpeg.org/ffmpeg-all.html#toc-replaygain), but inserting the proper tags/metadata into the file remains to be sorted (more reading required :) - that was why I asked the question. There's not *quite enough* detail in the Wikipedia article. 

And since Sonos can't deal with the ReplayGain tags, the other challenge is to find a streaming server that: 1.) will read and use the ReplayGain tags, and 2.) will work with Sonos. Plex meets the 2nd requirement, and it may meet the 1st requirement also - I've just not had time to confirm that. 

Anyway - that's where I am now. Again, your comments have been helpful, and I'd appreciate any others you care to make - as long as this isn't seen as too far off-topic here.

**Comment by pagdot** [comment]
@seamusdemora 
> Unfortunately, the Wikipedia article doesn't actually explain how to add the ReplayGain tags/metadata

It stored in the files metadata along artist, album, title, ... For MP3 this is the ID3 metadata. I'd recommend using a tool like [MP3TAG](https://www.mp3tag.de/en/) to explore/modify metadata. I'd recommend generating the replaygain values using e.g. ffmpeg and taking a look at them and editing them in MP3TAG. You will need to add the gain and peak columns.

**Comment by zrav** [comment]
You can also easily add RG tags in large batches using foobar2000. https://www.foobar2000.org/

**Comment by seamusdemora** [comment]
@pagdot 

> > Unfortunately, the Wikipedia article doesn't actually explain how to add the ReplayGain tags/metadata
> 
> It stored in the files metadata along artist, album, title, ... For MP3 this is the ID3 metadata. I'd recommend using a tool like [MP3TAG](https://www.mp3tag.de/en/) to explore/modify metadata. I'd recommend generating the replaygain values using e.g. ffmpeg and taking a look at them and editing them in MP3TAG. You will need to add the gain and peak columns.

I use `ffmpeg` and FLAC files - I have virtually no MP3 files in my library. Perhaps that is why terms like ID3 and MP3TAG don't mean anything to me. From what I've seen in the little time I've worked with `ffmpeg` and FLAC, an audio file contains 3 "sections": metadata (album title, artist, etc), an audio stream, and a video stream (which is actually a JPG/PNG album cover). 

I appreciate the recommendations. I will look into those if I can't discover how to do what I want in `ffmpeg`. I'm *under the impression* that `ffmpeg` is the "do-it-all" program, and so I am focused on that for now.

**Comment by seamusdemora** [comment]
@zrav:

> You can also easily add RG tags in large batches using foobar2000. https://www.foobar2000.org/

OK, that's an interesting suggestion - I will look into it - thanks. 

Question: You used the term *"in large batches"* wrt `foobar2000`, and that now has me wondering if I've confused something else. I thought that ReplayGain tags were ***relative*** to the "encoded volume" - that is, the ReplayGain tags are used to adjust the volume of an individual track (or album) by an amount needed to bring it to a desired value. 

If I were implanting tags in large batches, that suggests (to me at least) that I am setting a specific volume instead of normalizing a track by applying a gain to boost it to achieve a desired volume. In other words, RG tags would *tend to* be unique to each file (or perhaps an album). Have I misunderstood how RG works?

**Comment by zrav** [comment]
Yes, RG adds two tags to files, a track gain and an album gain (time-weighted average computed over all album tracks). I meant that in foobar2000 it is very easy to process many albums at once: Drag files to playlist, ctrl+A to select all, rightclick > ReplayGain > Scan as albums (by tags).
